[Remote Config] Add exponential backoff to HTTP client

ameske opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Draft PRs should be prefixed with `[WIP]` in their title.

-->
### What does this PR do?

Adds exponential backoff to the remote config HTTP client to
prevent agents from overwhelming the remote config edge during service
delivery issues.

### Motivation

The Remote Configuration backend will be responsible for serving
a large number of agents in the field. At the moment these agents
all poll Remote Config at a default polling interval of 1 minute.
Given the expected scope of the deployments, when the Remote Config
backend is unstable, it would be beneficial for agents to quickly
back their polling rates off until we recover.

### Additional Notes

This work leverages the existing `backoff` package already contained
within the agent codebase.

<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

If errors occur, agents could have a delay receiving updated config
once the service is stable. However, to counteract this the maximum
backoff interval is constrained to 5 minutes, and the recovery rate
is quick.

### Describe how to test/QA your changes

Unit tests are provided for how the backoff interval is calculated.

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).

Note: Adding GitHub labels is only possible for contributors with write access.
-->

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The appropriate `team/..` label has been applied, if known.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

ameske committed:
[Remote Config] Add exponential backoff to HTTP client

Adds exponential backoff to the remote config HTTP client to
prevent agents from overwhelming the remote config edge during service
delivery issues.

To do this, we leverage pkg/util/backoff and define a `backoff.Policy`
for the remote config client. We keep track of the current error level
on successful and failed updates, and calculate a refresh interval for
each request to the remote config backend based on the state of the
system.

The backoff policy has jitter built in for random backoff times. The
default values we use for the policy start with a backoff duration that
will be within the range [30s, 60s], with the min and max doubling with
every consecutive error. When we see a success, we drop back two
states. (equivalent to dividing the min and the max by 4) The maximum
backoff added is is configurable and is constrained to be between 2m
and 5m. No other values related to the backoff policy are configurable
by the user at this time.

Resolves: [RCM-58]

jtappa reviewed (commented):
👍🏻 for docs/release notes.

hush-hush commented on pkg/config/remote/service/service_test.go:
```diff
@@ -84,6 +86,93 @@ func newTestService(t *testing.T, api *mockAPI, uptane *mockUptane, clock clock.
+	assert.GreaterOrEqual(t, refreshInterval, 121*time.Second)
+	assert.LessOrEqual(t, refreshInterval, 241*time.Second)
+
+	// Make the call succeed so we can test the recovery interval
+	api = &mockAPI{}
+	api.On("Fetch", mock.Anything, &pbgo.LatestConfigsRequest{
+		Hostname:                     service.hostname,
+		AgentVersion:                 version.AgentVersion,
+		CurrentConfigSnapshotVersion: 0,
+		CurrentConfigRootVersion:     0,
+		CurrentDirectorRootVersion:   0,
+		Products:                     []string{},
+		NewProducts:                  []string{},
+	}).Return(lastConfigResponse, nil)
+	service.api = api
```
This seems like a second test case. Could we split this test into more "unit" like one ?

ameske replied:
Yep, can do!

hush-hush commented on releasenotes/notes/remote-config-http-client-backoff-d9acbc7bc24b4ca7.yaml:
```diff
@@ -0,0 +1,13 @@
+#
+# Each section note must be formatted as reStructuredText.
+---
+enhancements:
+  - |
+    The Remote Configuration client will do exponential backoff
+    for subsequent requests when errors are encountered to preserve
+    service stability.
```
Is remote configuration public ? I could not find anything in our release note and the configuration isn't in the config template (which make me this it's still internal/beta)

Release notes are used to generate the CHANGELOG which is meant for users not dev. We add item to it only if we want any user/customer to use the new feature. I guess this PR should not have a release note. You can use the `changelog/no-changelog` Github label to pass the CI.

ameske replied:
It's still only internal.

ameske committed:
Remove release notes

Remote config is not external at this time and thus we don't need
release notes to let the user know about this change.

ameske committed:
Make testing error recovery a separate test

ameske committed:
Explicitly state we are overriding config values

ameske committed:
CircleCI judged me for my spelling skills
